src/qibo/backends/numpy.py

```diff
@@ -479,55 +507,64 @@ def execute_circuit_repeated(self, circuit, initial_state=None, nshots=None):
                             gate.substitute_symbols()
                         state = gate.apply(self, state, nqubits)
 
+            if circuit.density_matrix:
+                final_states.append(state)
             if circuit.measurements:
-                result = CircuitResult(self, circuit, state, 1)
+                result = CircuitResult(state, circuit.measurements, self, 1)
                 sample = result.samples()[0]
                 results.append(sample)
                 if not circuit.density_matrix:
-                    probabilities += result.probabilities()
+                    probabilities += result.probabilities(target_qubits)
                     samples.append("".join([str(s) for s in sample]))
-            else:
-                results.append(state)
 
-        if circuit.measurements:
-            final_result = CircuitResult(self, circuit, state, nshots)
-            final_result._samples = self.aggregate_shots(results)
-            if not circuit.density_matrix:
-                final_result._repeated_execution_probabilities = probabilities / nshots
-                final_result._repeated_execution_frequencies = (
-                    self.calculate_frequencies(samples)
+        if circuit.density_matrix:  # this implies also it has_collapse
+            assert circuit.has_collapse
+            final_state = np.asarray(final_states).mean(0)
+            if circuit.measurements:
+                qubits = [q for m in circuit.measurements for q in m.target_qubits]
+                final_result = CircuitResult(
+                    final_state, circuit.measurements, self, nshots
                 )
+                probabilities = final_result.probabilities(qubits)
+                final_result._repeated_execution_probabilities = probabilities
+                circuit._final_state = final_result
+                return final_result
+            else:
+                final_result = QuantumState(final_state, self)
+                probabilities = final_result.probabilities()
+                final_result._repeated_execution_probabilities = probabilities
+                circuit._final_state = final_result
+                return final_result
+        else:
+            probabilities = probabilities / nshots
+            final_result = MeasurementOutcomes(
+                circuit.measurements, probabilities, self, nshots
+            )
+            final_result._samples = self.aggregate_shots(results)
+            final_result._repeated_execution_probabilities = probabilities
```

Comment:
Also here you are assigning the `_repeated_execution_probabilities` attribute to a `MeasurementOutcomes` object, but this does not exist in the original class definition. Python allows you to set attributes to instantiated objects, even if they do not exist in their definition but it's not a very good practice because then it's harder to track what attributes each object has. Pylint usually complains for this, I am not sure why it is passing now.

If the `_repeated_execution_probabilities` is not used we can remove, otherwise we should define it as `None` in the `MeasurementOutcomes.__init__`.